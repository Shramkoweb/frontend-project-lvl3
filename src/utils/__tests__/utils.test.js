import path from 'path';
import fs from 'fs';

import {
  createElement,
  removeTrailingSlashes,
  render,
  renderPositions,
  validate,
} from '../utils';

const getFixturePath = (filename) => path.join(process.cwd(), '/src', '/__fixtures__', filename);
const readFixtureContent = (filename) => fs.readFileSync(getFixturePath(filename), 'utf-8').trim();

describe('createElement', () => {
  let html;

  beforeAll(() => {
    html = readFixtureContent('test.html');
  });

  it('should return Node', () => {
    const element = createElement(html);

    expect(element).toBeInstanceOf(Node);
  });

  it('should render correct markup', () => {
    const element = createElement(html);

    expect(element).toMatchSnapshot();
  });

  it('should throw error without template', () => {
    const callWithError = () => {
      createElement();
    };

    expect(callWithError).toThrowErrorMatchingSnapshot();
  });
});

describe('render', () => {
  let container;
  let testElement;
  let html;

  beforeAll(() => {
    html = readFixtureContent('test.html');
    testElement = document.createElement('p');
    testElement.textContent = 'Render test element';
  });

  beforeEach(() => {
    container = createElement(html);
  });

  afterEach(() => {
    container = null;
  });

  it('should throw error without place', () => {
    const callWithError = () => {
      render();
    };

    expect(callWithError).toThrowErrorMatchingSnapshot();
  });

  it('should render AFTERBEGIN ', () => {
    render(container, testElement, renderPositions.AFTERBEGIN);

    expect(container).toMatchSnapshot();
  });

  it('should render BEFOREEND ', () => {
    render(container, testElement, renderPositions.BEFOREEND);

    expect(container).toMatchSnapshot();
  });
});

describe('removeTrailingSlashes', () => {
  it('should remove traling slashes', () => {
    const current = removeTrailingSlashes('https://css-tricks.com/feed//');
    const expected = 'https://css-tricks.com/feed';

    expect(current).toEqual(expected);
  });
});

describe('validate', () => {
  it('should return URL if valid', () => {
    const url = 'https://css-tricks.com/feed';

    return validate(url)
      .then((data) => expect(data)
        .toBe(url));
  });

  it('should reject with error if invalid URL', () => {
    const invalidURL = 'https//css-tricks.com/feed';
    const valid = validate(invalidURL);

    return expect(valid).rejects.toMatchSnapshot();
  });

  it('should reject with error if empty URL', () => {
    const invalidURL = '';
    const valid = validate(invalidURL);

    return expect(valid).rejects.toMatchSnapshot();
  });

  it('should reject with error if URL already have', () => {
    const url = 'https://css-tricks.com/feed';
    const valid = validate(url, [url]);

    return expect(valid).rejects.toMatchSnapshot();
  });
});
