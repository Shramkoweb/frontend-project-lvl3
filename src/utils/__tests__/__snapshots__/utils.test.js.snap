// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createElement should render correct markup 1`] = `
<div
  class="container"
>
  
  
  <h1
    class="title"
  >
    Test title
  </h1>
  
  
  <p
    class="description"
  >
    Test description
  </p>
  

</div>
`;

exports[`createElement should throw error without template 1`] = `"Unexpected call createElement without template"`;

exports[`render should render AFTERBEGIN  1`] = `
<div
  class="container"
>
  <p>
    Render test element
  </p>
  
  
  <h1
    class="title"
  >
    Test title
  </h1>
  
  
  <p
    class="description"
  >
    Test description
  </p>
  

</div>
`;

exports[`render should render BEFOREEND  1`] = `
<div
  class="container"
>
  
  
  <h1
    class="title"
  >
    Test title
  </h1>
  
  
  <p
    class="description"
  >
    Test description
  </p>
  

  <p>
    Render test element
  </p>
</div>
`;

exports[`render should throw error without place 1`] = `"Uncorrected render place undefined"`;

exports[`validate should reject with error if URL already have 1`] = `[ValidationError: hasUrlYet]`;

exports[`validate should reject with error if empty URL 1`] = `[ValidationError: emptyString]`;

exports[`validate should reject with error if invalid URL 1`] = `[ValidationError: invalidUrl]`;
